src/servers/src/interceptor.rs

```diff
@@ -397,3 +398,61 @@ impl<E: ErrorExt> PromStoreProtocolInterceptor for Option<PromStoreProtocolInter
         }
     }
 }
+
+/// LogIngestInterceptor can track life cycle of a log ingestion request
+/// and customize or abort its execution at given point.
+pub trait LogIngestInterceptor {
+    type Error: ErrorExt;
+
+    /// Called before pipeline execution.
+    fn pre_pipeline(
+        &self,
+        _values: &[Value],
+        _query_ctx: QueryContextRef,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    /// Called before insertion.
+    fn pre_ingest(
+        &self,
+        _request: &RowInsertRequests,
+        _query_ctx: QueryContextRef,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
```

Comment:
Not sure if it's fully possible but if we can take values and request, then return them from interceptor, it can be great. This will allow us to modify the payload of request.